### PR TITLE
ISPN-1149 Allow the protocol servers to choose their own port

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-1149

### CODE REVIEW

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/test/HotRodClientTestingUtil.java [hunk 1]
```diff
@@ -48,7 +48,7 @@
    private static final int DEFAULT_PORT = 15232;
 
    public static HotRodServer startHotRodServer(EmbeddedCacheManager cacheManager, HotRodServerConfigurationBuilder builder) {
-      return startHotRodServer(cacheManager, findFreePort(), builder);
```
Comment (slaskawi) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/test/HotRodClientTestingUtil.java]: Perhaps we can remove `#findFreePort` method completely?

Comment (gustavonalle) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/test/HotRodClientTestingUtil.java]: I'd rather keep it; it is often needed for tests

File: server/core/src/main/java/org/infinispan/server/core/AbstractProtocolServer.java [hunk 2]
```diff
@@ -83,7 +85,15 @@ public final void start(A configuration, EmbeddedCacheManager cacheManager) {
    }
 
    protected void startTransport() {
-      InetSocketAddress address = new InetSocketAddress(configuration.host(), configuration.port());
+      int port = configuration.port();
```
Comment (pruivo) [inline server/core/src/main/java/org/infinispan/server/core/AbstractProtocolServer.java]: I don't think that this code is needed. From the `InetSocketAddress` javadoc:
```
 A port number of zero will let the system pick up an ephemeral port in a bind operation
```
https://docs.oracle.com/javase/8/docs/api/java/net/InetSocketAddress.html#InetSocketAddress-java.net.InetAddress-int-

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/configuration/HotRodServerConfigurationBuilder.java [hunk 3]
```diff
@@ -125,6 +125,9 @@ public void validate() {
       if (proxyHost == null && host == null) {
          throw log.missingHostAddress();
       }
+      if (proxyPort < 0 && port == 0 && !startTransport) {
```
Comment (pruivo) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/configuration/HotRodServerConfigurationBuilder.java]: why are you checking the `proxyPort`? I don't see why you can't have a `proxyPort` with a "ephemeral port".

Comment (tristantarrant) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/configuration/HotRodServerConfigurationBuilder.java]: Right, I need to think about this a bit more

Comment (danberindei) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/configuration/HotRodServerConfigurationBuilder.java]: @pruivo The (reverse) proxy has to know about the server's port in order to forward requests, I'm not sure how that would work if the server port changes on each restart.

**Comment by tristantarrant** [comment]
Temporarily closing because of CI overload

**Pull request closed without merging**